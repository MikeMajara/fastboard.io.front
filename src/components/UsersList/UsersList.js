import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { formatDate } from '#utils';

import Button from '../Button';

import {
  SCard, SHeader, SContent, SUser, SUserJoinDate, SFooter,
} from './styled';

// @todo scroll when list is too long
const UsersList = ({
  boardName, myUserName, myJoinDate, others, onClose,
}) => {
  const { t } = useTranslation('board');

  const myJoinDateString = useMemo(
    () => formatDate(new Date(myJoinDate)),
    [myJoinDate],
  );

  const othersList = useMemo(
    () => others.map(({ id, name, joinDate }) => ({
      id,
      name,
      date : formatDate(new Date(joinDate)),
    })),
    [others],
  );

  return (
    <SCard>
      <SHeader>
        {t('meta.usersList.title', {
          count : 1 + others.length,
          boardName,
        })}
      </SHeader>
      <SContent>
        <SUser>
          {t('meta.usersList.you', { name : myUserName })}
          <SUserJoinDate>
            {t('meta.usersList.youJoined', { date : myJoinDateString })}
          </SUserJoinDate>
        </SUser>
        {othersList.map(({ id, name, date }) => (
          <SUser key={id}>
            {name}
            <SUserJoinDate>
              {t('meta.usersList.otherJoined', { date })}
            </SUserJoinDate>
          </SUser>
        ))}
      </SContent>
      <SFooter>
        <Button type="secondary" onClick={onClose}>
          {t('meta.usersList.accept')}
        </Button>
      </SFooter>
    </SCard>
  );
};

UsersList.propTypes = {
  boardName  : PropTypes.string.isRequired,
  myUserName : PropTypes.string.isRequired,
  myJoinDate : PropTypes.string.isRequired,
  others     : PropTypes.arrayOf(PropTypes.shape({
    id       : PropTypes.string.isRequired,
    name     : PropTypes.string.isRequired,
    joinDate : PropTypes.string.isRequired,
  })).isRequired,
  onClose : PropTypes.func.isRequired,
};

export default UsersList;
