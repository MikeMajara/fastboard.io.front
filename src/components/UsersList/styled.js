import styled, { css, down } from '@xstyled/styled-components';

import Card from '../Card';

export const SCard = styled(Card)`
  display: flex;
  flex-direction: column;
  justify-content: center;
  ${down('md', css`text-align: center;`)}
`;

export const SHeader = styled.h2`
  margin: 0;
  font-size: xlg;
`;

export const SContent = styled.ul`
  padding: 0;
  margin: default 0;
  list-style: none;
`;

export const SUser = styled.li`
  display: flex;
  justify-content: space-between;

  & + & {
    margin-top: default;
  }
`;

export const SUserJoinDate = styled.span`
  color: primary;
`;

export const SFooter = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;

  ${down('md', css`button { width: 100%; }`)}
`;
