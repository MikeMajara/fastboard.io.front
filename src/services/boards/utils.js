/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
