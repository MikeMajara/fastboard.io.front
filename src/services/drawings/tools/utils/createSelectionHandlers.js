import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
