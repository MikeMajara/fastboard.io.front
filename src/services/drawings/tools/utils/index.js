export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
