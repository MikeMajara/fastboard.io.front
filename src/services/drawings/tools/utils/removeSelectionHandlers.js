import { Log } from '#utils';

/**
 * Remove selection handlers.
 *
 * @param {Object} selectedItem Item to be removed selection handlers from.
 * @param {Object} selectionLayer Layer that contains the selection handlers.
 */
export default function removeSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : removeSelectionHandlers');

  selectedItem.set({ selected : false });
  selectionLayer.removeChildren();
}
