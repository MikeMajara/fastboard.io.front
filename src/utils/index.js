export { default as Log } from './logger';
export { default as generateId } from './generateId';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as validBoardId } from './validBoardId';
