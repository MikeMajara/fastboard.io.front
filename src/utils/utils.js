/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
