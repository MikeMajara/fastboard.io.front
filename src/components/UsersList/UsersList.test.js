import React from 'react';
import { fireEvent } from '@testing-library/react';

import { render } from '#devTools/tests/utils';

import UsersList from './UsersList';

const boardName = 'My board';
const myUserName = 'My user name';
const myJoinDate = '2021-02-28T12:14:15.000Z';
const others = [
  {
    id       : '1',
    name     : 'other user 1',
    joinDate : '2021-02-28T12:15:00.000Z',
  },
  {
    id       : '2',
    name     : 'other user 2',
    joinDate : '2021-02-28T12:15:00.000Z',
  },
];

describe('Component : UsersList', () => {
  test('Renders users list correctly and reacts to close action', () => {
    const onCloseFn = jest.fn();
    const { getByText } = render(
      <UsersList
        boardName={boardName}
        myUserName={myUserName}
        myJoinDate={myJoinDate}
        others={others}
        onClose={onCloseFn}
      />,
    );

    expect(getByText('meta.usersList.title')).toBeInTheDocument();
    expect(getByText(others[0].name)).toBeInTheDocument();
    expect(getByText(others[1].name)).toBeInTheDocument();
    fireEvent.click(getByText('meta.usersList.accept'));
    expect(onCloseFn).toHaveBeenCalledTimes(1);
  });

  test('Snapshot', () => {
    const { asFragment } = render(
      <UsersList
        boardName={boardName}
        myUserName={myUserName}
        myJoinDate={myJoinDate}
        others={others}
        onClose={jest.fn()}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
