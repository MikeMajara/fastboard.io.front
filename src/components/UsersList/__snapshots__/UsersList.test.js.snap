// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : UsersList Snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-qZtCU XqBnS sc-pzMyG gEsWbO"
  >
    <h2
      class="sc-pHIBf hewvQA"
    >
      meta.usersList.title
    </h2>
    <ul
      class="sc-pQEbo iICRvb"
    >
      <li
        class="sc-pYA-dN bzRDoT"
      >
        meta.usersList.you
        <span
          class="sc-oUcyK bWYQaI"
        >
          meta.usersList.youJoined
        </span>
      </li>
      <li
        class="sc-pYA-dN bzRDoT"
      >
        other user 1
        <span
          class="sc-oUcyK bWYQaI"
        >
          meta.usersList.otherJoined
        </span>
      </li>
      <li
        class="sc-pYA-dN bzRDoT"
      >
        other user 2
        <span
          class="sc-oUcyK bWYQaI"
        >
          meta.usersList.otherJoined
        </span>
      </li>
    </ul>
    <div
      class="sc-pbYBj jxvTZU"
    >
      <button
        class="sc-qQYBZ kBBHFN"
        type="secondary"
      >
        meta.usersList.accept
      </button>
    </div>
  </div>
</DocumentFragment>
`;
